File: SampleObjC/sdk-ios-objc/ViewController.m
//
//  ViewController.m
//  sdk-ios-objc
//
//  Created by Simon Hamadene on 16/10/2018.
//  Copyright © 2018 Yoti Limited. All rights reserved.
//

#import "ViewController.h"
#import <YotiButtonSDK/YotiButtonSDK.h>
#import "ProfileViewController.h"

@interface ViewController () <YTBScenarioRetrievalDelegate, YTBBackendDelegate>
@property (weak, nonatomic) IBOutlet YotiButton *rememberMeButton;
@property (weak, nonatomic) IBOutlet YotiButton *selfieAuthButton;

@end

@implementation ViewController {
    NSDictionary *responseObject;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.rememberMeButton setTitle:@"RememberMe Scenario" forState: UIControlStateNormal ];
    [self.selfieAuthButton setTitle:@"SelfieAuth Scenario" forState: UIControlStateNormal ];
}

-(void) moveToProfile {
    [self performSegueWithIdentifier:@"moveToProfile" sender:self];
}

-(void) prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    
    if ([[segue identifier] isEqualToString:@"moveToProfile"])
    {
        ProfileViewController* profileViewController = [segue destinationViewController];
        
        NSData* selfieValue = [responseObject objectForKey:@"selfie"];
        if (selfieValue != nil) {
            NSData* imageData = [[NSData alloc] initWithData:selfieValue];
            profileViewController.selfie = [[UIImage alloc] initWithData:imageData];
        }
        NSString* phoneNumberValue = [responseObject objectForKey:@"phone_number"];
        if ( ![phoneNumberValue isEqual:@""]) {
            profileViewController.phone = phoneNumberValue;
        }
        NSString* givenNamesValue = [responseObject objectForKey:@"given_names"];
        if ( ![givenNamesValue isEqual:@""]) {
            profileViewController.givenNames = givenNamesValue;
        }
        NSString* postalAddressValue = [responseObject objectForKey:@"postal_address"];
        if ( ![postalAddressValue isEqual:@""]) {
            profileViewController.postalAddress = postalAddressValue;
        }
        NSString* genderValue = [responseObject objectForKey:@"gender"];
        if ( ![genderValue isEqual:@""]) {
            profileViewController.gender = genderValue;
        }
        NSString* emailAddressValue = [responseObject objectForKey:@"email_address"];
        if ( ![emailAddressValue isEqual:@""]) {
            profileViewController.emailAddress = emailAddressValue;
        }
        NSString* familyNameValue = [responseObject objectForKey:@"family_name"];
        if ( ![familyNameValue isEqual:@""]) {
            profileViewController.familyName = familyNameValue;
        }
        NSString* dateOfBirthValue = [responseObject objectForKey:@"date_of_birth"];
        if ( ![dateOfBirthValue isEqual:@""]) {
            profileViewController.dateOfBirth = dateOfBirthValue;
        }
    }
    
}

- (void)yotiSDKDidFailFor:(NSString * _Nonnull)useCaseID with:(NSError * _Nonnull)error {
}

- (void)yotiSDKDidSucceedFor:(NSString * _Nonnull)useCaseID baseURL:(NSURL * _Nullable)baseURL token:(NSString * _Nullable)token url:(NSURL * _Nullable)url {
    YTBScenario *scenario = [YotiSDK scenarioforUseCaseID:useCaseID];
    
    [YotiSDK callbackBackendScenario:scenario token:token withDelegate:self];
}

- (void)backendDidFinishWith:(NSData * _Nullable)data error:(NSError * _Nullable)error {
    
    if (data != nil) {
        NSError *jsonError = nil;
        
        responseObject = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonError];
    }
    
}


@end

File: Sample/sdk-ios-swift/ViewController.swift
//
//  Copyright © 2018 Yoti Ltd. All rights reserved.
//

import UIKit
import YotiButtonSDK

class ViewController: UIViewController {

    var responseObject: ProfileDictionary?
    @IBOutlet weak var rememberMeButton: YotiButton!
    @IBOutlet weak var selfieAuthButton: YotiButton!
    @IBOutlet weak var unfulfilledButton: YotiButton!

    func yotiButtonDidTouchUpInside(_ sender: YotiButton) {
        guard let useCaseID = sender.useCaseID else {
             return
        }
        do {
            UIApplication.shared.isNetworkActivityIndicatorVisible = true
            try YotiSDK.startScenario(for: useCaseID, theme: sender.theme, with: self)
        } catch let error {
            print("\(error.localizedDescription)")
        }
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        rememberMeButton.theme = .yoti
        unfulfilledButton.theme = .easyID

        let action: YotiButton.TouchedUpInside = { [weak self] button in
            self?.yotiButtonDidTouchUpInside(button)
        }

        rememberMeButton.action = action
        selfieAuthButton.action = action
        unfulfilledButton.action = action
    }

    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        rememberMeButton.setTitle("RememberMe Scenario", for: .normal)
        selfieAuthButton.setTitle("SelfieAuth Scenario", for: .normal)
    }

    func moveToProfile() {
        performSegue(withIdentifier: "moveToProfile", sender: self)
    }

    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        guard let identifier = segue.identifier else { return }
        switch identifier {
        case "moveToProfile":
            guard let viewController = segue.destination as? ProfileViewController else { break }
            guard let responseObject = responseObject, !responseObject.attributes.isEmpty else { break }
            responseObject.attributes.forEach {

                if $0.name == "selfie" {
                    let selfieValue = $0.value
                    if let imageData = Data(base64Encoded: selfieValue) {
                        let photo = UIImage(data: imageData)
                        viewController.selfie = photo
                    }
                }
                if $0.name == "phone_number" {
                    let phoneNumberValue = $0.value
                    viewController.phone = phoneNumberValue
                }
                if $0.name == "given_names"{
                    let givenNamesValue = $0.value
                    viewController.givenNames = givenNamesValue
                }
                if $0.name == "postal_address"{
                    let postalAddressValue = $0.value
                    viewController.postalAddress = postalAddressValue
                }
                if $0.name == "gender"{
                    let genderValue = $0.value
                    viewController.gender = genderValue
                }
                if $0.name == "email_address"{
                    let emailAddressValue = $0.value
                    viewController.emailAddress = emailAddressValue
                }
                if $0.name == "family_name"{
                    let familyNameValue = $0.value
                    viewController.familyName = familyNameValue
                }
                if $0.name == "date_of_birth"{
                    let dateOfBirthValue = $0.value
                    viewController.dateOfBirth = dateOfBirthValue
                }
            }
        default:
            break
        }
    }
}

extension ViewController: SDKDelegate {
    func yotiSDKDidFail(for useCaseID: String, appStoreURL: URL?, with error: Error) {
        UIApplication.shared.isNetworkActivityIndicatorVisible = false
        print(error)
    }

    func yotiSDKDidSucceed(for useCaseID: String, baseURL: URL?, token: String?, url: URL?) {
        let scenario = YotiSDK.scenario(for: useCaseID)
        YotiSDK.callbackBackend(scenario: scenario!, token: token!, with: self)
    }

    func yotiSDKDidOpenYotiApp() {
        UIApplication.shared.isNetworkActivityIndicatorVisible = false
    }
}

extension ViewController: BackendDelegate {
    func backendDidFinish(with data: Data?, error: Error?) {
        guard let data = data else {
            return
        }
        do {
            let decodedJson = try JSONDecoder().decode(ProfileDictionary.self, from: data)
            responseObject = decodedJson
            moveToProfile()
        } catch let error {
            print (error)
        }
    }
}
